### ISSUE TITLE
Provide information in the API about what contacts should have been provided

### ISSUE THREAD

**Issue by ColemanCollins** [5calls/5calls#176]
This is related to #64 and #175 (and maybe even other things like #171)

It would be helpful to have some way to understand on the front-end what contacts _should have been provided_ based on how they're categorized in airtable (ie: this issue is relevant to your reps in the Senate, House, Both). Currently, we only see what the API has returned to us for contacts with no understanding if this is the complete intended set of contacts.

This would let us provide issue-level messaging around split districts (eg: "you also have a house rep to call about this but we can't show it to you unless you provide a better location") and better general messaging around Civic API freakouts ("we're providing no contacts for this even though we should be showing you your senators and we have no idea why").

**Comment by nickoneill** [comment]
Some of this is starting to come back in the API. There's a `splitDistrict` bool in the root and some errors for invalid addresses as well.

What else specifically would be helpful? I'm not sure how to detect civic api freakouts, though I suspect they do happen. Any more info here would be nice.

**Comment by nickoneill** [comment]
I updated the comment above after reviewing the other PR you sent ^

**Comment by oderby** [comment]
I think the point here is that the issue scripts in the db indicate "this applies to just senators" vs "this issue applies to just representatives" vs "this applies to both", but the backend processes that and strips it out before sending the script + a list of contacts to the client, leaving no indication to the client what the expected number of contacts was. Given that we only support 3 scenarios [2 senators | 1 rep | 2 senators + 1 rep], the length of contacts list and splitDistrict boolean is _almost_ sufficient to determine if and what is missing:

1. `contacts.length === 3`: We have all the contacts! (or something is really wrong...)
2. `contacts.length === 2 && !splitDistrict`: We have all the contacts (must be just senators)
3. `contacts.length === 2 && splitDistrict`: We're missing your rep - try to give more information (*or* civic api freakout, sorry...)
4. `contacts.length === 1`: We have your rep (and that's all we expect)
5. `contacts.length === 0`: uhh...??? (Something's really wrong, shouldn't happen...)

Now, 3 is the only somewhat ambiguous case, but as @nickoneill points out, the backend doesn't really know that it's a civic api freakout vs "there's some more specific address you could give to get the right rep" (because we just [look at the response](https://github.com/5calls/5calls/blob/2152ceca4d1cb2e3dd16e2560ef3ae6bb34f8109/go/handler.go#L56) to see if a rep is returned or not). Unfortunately it's likely not very ambiguous to our users - they've already seen the error message and went ahead and put in a specific address, and now you're just showing them the same thing. So we should update the error message to indicate that it could be an issue with the system...and maybe encourage them to report it?

But all the above makes the big assumption that our issues will only ever target reps and senators, which I presume isn't true. So while `splitDistrict` solves this issue for the moment, we should probably figure out a better way to indicate this discrepancy in the future...

**Comment by ColemanCollins** [comment]
@oderby `splitDistrict` alone actually doesn't solve the current issue either; a bit more information is required even in that case, because the `splitDistrict` flag is something that is just being passed through from Google Civic and while it impacts the contacts provided in some cases, it doesn't in others, so even though we know `splitDistrict` and `contacts.length` at whatever the API generated, we still don't know from the front-end the total number of _intended_ contacts.

Two specific examples:
1. `contacts.length === 2 && splitDistrict` — in this scenario, for a given issue, it's possible that  we're missing your rep and we've only returned 2 of a possible 3 contacts OR we only ever intended to return your senators (so 2 of 2), and it just so happens that your location is in a split district. Because `splitDistrict` _may or may not_ impact the response based on the intent, we're not sure which of these cases is true.
1. `contacts.length === 0 && splitDistrict` — this is the scenario that PR #175 covers and even it's not actually perfect, because it doesn't take into account the possibility of strangeness emanating from Civic API (see: #171). Still better better than just showing a "I dunno" message because all of the other cases that could get you to that scenario are problems that you can't fix as a caller, but not perfect.

There's a non-technical (and/or short-term) solution to this problem by making `splitDistrict` messaging that's unrelated to the issue at hand in non-no-contact scenarios e.g. "hey your address isn't specific enough to show you your house rep (or just complete list of contacts, in the case that that's eventually a thing)", without specifically messaging it as "this is stopping us from showing you someone to call for _this issue_." 

But IMO it would be superior in terms of callers understanding what's going on to provide that message contextually ("you have one more person left to call for this issue; if you give us a more specific location we'll also add your house rep to that list"), and not present it as a problem at all if it's not relevant to the specific issue. 

That is to say, in the interest of simplicity and bias-towards-action on the part of the callers (don't make them solve a problem they don't yet have), I'd like to have the knowledge to show split district messaging where and only where it would impact the caller's experience or workflow, and the knowledge to hide it in cases where it's still true but doesn't matter.

**Comment by ColemanCollins** [comment]
My idea behind _intended contacts_ as a solution is it reduces the complexity of the error cases — instead of comparing a bunch of parameters and inferring what went wrong, it becomes a simple two-parter:

We have returned all of the contacts we intended to return for this issue (yay!) OR we have not returned all of the contacts we intended to for this issue (oh no!)

GIVEN we have not returned all of the contacts we intended to for this issue, we either don't have enough information to return all of the contacts we intended to return (`splitDistrict`) OR something else went wrong (that the caller probably can't impact)

Intended contacts could probably just be a number— "we expected 3" or specifics "we expected Senators and Reps." In either case, the front end would have to know/infer some things about there being 2 senators and 1 rep for a given congressional district. Could also be something else entirely. I'm extremely not an expert in API design.

**Comment by oderby** [comment]
> That is to say, in the interest of simplicity and bias-towards-action on the part of the callers (don't make them solve a problem they don't yet have), I'd like to have the knowledge to show split district messaging where and only where it would impact the caller's experience or workflow, and the knowledge to hide it in cases where it's still true but doesn't matter.

I agree 100% with this intention. This is what we should be striving towards. One caveat/clarification - I think we should always give an explanation/indication re:expected vs available contacts, even if it's not something the user can change. I know I'd lose confidence in 5calls if I expected to be calling my rep in addition to senators but my rep wasn't listed and there was no explanation for why.

Longer term, I don't think just a number of `intendedContacts` will be sufficient - as we add governors and state legislators, I can imagine it's not clear who's missing vs who's expected. And I think we want as much of this logic in the backend, rather than duplicated across our 3 (so far) different clients.

A naive proposal is that the returned contacts list always has an entry for every expected contact (removing the confusion around how many there should be). Then, for missing contacts we provide "dummy/missing" entries, which specify what we do know (e.g. that it would be your representative, but we couldn't look it up because of split district) but leave the other field blank.

**Comment by ColemanCollins** [comment]
This makes sense to me. Happy to pair with you on this from a front-end/design perspective if you feel confident proposing/implementing the backend changes.  A little concerned about creating a breaking change in the API with a feature like this — I assume @nickoneill can give us guidance on that.

**Comment by oderby** [comment]
Yeah, I'm game. Need to gather some additional details on the schema/templating for issues/scripts and how details such as targeted MoC are specified before implementing, but I can start on a more formalized proposal of the above (@nickoneill feel free to chime in if you disagree on priority or approach).

**Comment by oderby** [comment]
Okay, so I've realized that my preferred fix would be a breaking change to the GET /issues/ API (changing behavior of the `contacts` field in the case of splitDistrict/civic api failures), which might not be desirable for our mobile clients. So I've outlined two variations of the proposal below. Any guidance/preferences/thoughts are welcome. 

**Proposal 1: non-breaking fix** Backend adds a new field to the `Issue` struct, called `missingContacts`, which is of type `[]Contact` and will be populated with a list of contacts which the issue expects but we were unable to retrieve from the Civic API (due to [splitDistrict](https://github.com/5calls/5calls/issues/64), error or [civic api freakout](https://github.com/5calls/5calls/issues/171)). These contacts will only have the `area` field specified. Otherwise, leave all existing fields of `Issue` struct as-is.

**Proposal 2: breaking change** Backend changes semantics of the `contacts` field of the `Issue` struct to include entries for all expected contacts, regardless of whether we are able to fetch contact information for them or not. For contacts that we are not able to retrieve from Civic API (for same reasons as above), we treat them the same as above (the only field filled out is the `area` field).

So, here's a contrived example in terms of the JSON response (contrived because it will exhibit many of the cases outlined above in one response, which is almost certainly not possible)
_Current_
```javascript
{
    "splitDistrict":true,
    "invalidAddress":false,
    "normalizedLocation":"North Palm Beach",
    "issues":[
        // expected 2 senators, but no contacts found, due to #171
        {"id":"1",...,"contacts":[]}, 
        // expected 1 Rep, but no contacts found, due to splitDistrict
        {"id":"2",...,"contacts":[]}, 
        // expected 2 senators, got 2 senators
        {"id":"3",...,"contacts":[    
            {"id":"FL-BillNelson",...,"area":"Senate"},
            {"id":"FL-MarcoRubio",...,"area":"Senate"}
        ]},
        // expected 2 senators + 1 rep, got just 2 senators due to splitDistrict
        {"id":"4",...,"contacts":[    
            {"id":"FL-BillNelson",...,"area":"Senate"},
            {"id":"FL-MarcoRubio",...,"area":"Senate"}
        ]}
    ]
}
```
_Response with Proposal 1_
```javascript
{
    "splitDistrict":true,
    "invalidAddress":false,
    "normalizedLocation":"North Palm Beach",
    "issues":[
        // expected 2 senators, but no contacts found, due to #171
        {"id":"1",...,"contacts":[], "missingContacts":[{"area":"Senate"}, {"area":"Senate"}]},
        // expected 1 Rep, but no contacts found, due to splitDistrict
        {"id":"2",...,"contacts":[], "missingContacts":[{"area":"House"}]},
        // expected 2 senators, got 2 senators
        {"id":"3",...,"contacts":[
            {"id":"FL-BillNelson",...,"area":"Senate"},
            {"id":"FL-MarcoRubio",...,"area":"Senate"}
        ], "missingContacts":[]},
        // expected 2 senators + 1 rep, got just 2 senators due to splitDistrict
        {"id":"4",...,"contacts":[
            {"id":"FL-BillNelson",...,"area":"Senate"},
            {"id":"FL-MarcoRubio",...,"area":"Senate"}
        ], "missingContacts":[{"area":"House"}]}
    ]
}
```
_Response with Proposal 2_
```javascript
{
    "splitDistrict":true,
    "invalidAddress":false,
    "normalizedLocation":"North Palm Beach",
    "issues":[
        // expected 2 senators, but no contacts found, due to #171
        {"id":"1",...,"contacts":[{"area":"Senate"}, {"area":"Senate"}]},
        // expected 1 Rep, but no contacts found, due to splitDistrict
        {"id":"2",...,"contacts":[{"area":"House"}]},
        // expected 2 senators, got 2 senators
        {"id":"3",...,"contacts":[
            {"id":"FL-BillNelson",...,"area":"Senate"},
            {"id":"FL-MarcoRubio",...,"area":"Senate"}
        ]},
        // expected 2 senators + 1 rep, got just 2 senators due to splitDistrict
        {"id":"4",...,"contacts":[
            {"id":"FL-BillNelson",...,"area":"Senate"},
            {"id":"FL-MarcoRubio",...,"area":"Senate"},
            {"area":"House"}
        ]}
    ]
}
```

**Comment by nickoneill** [comment]
I'm more fond of proposal 1, not just because it doesn't break existing things but it missing contacts feel like something that could be ignored in some cases which makes a bit more sense in 1 than 2.

Before diving too far in here: there is some new messaging in place to upsell split districts to enter an address. What's the frontend for this look like and does it meaningfully improve the experience for users? I understand we can't perfectly know the contacts that are missing in the current form but we can always reasonably say "you should enter a more specific address" (except the case of civic API freakout which I'm not certain happens with any significance).

The process by which you've come up with a solution is great, just measuring the amount of impact this will have as a quick check. Sorry for not finding this earlier, things needing attention are coming from lots of directions right now.

**Comment by oderby** [comment]
No worries - totally understand there's a lot to go through ;) And good point to take step back. I think if we're not confident this is high value, we could start to answer it by measuring how many people are in split districts and how many are impacted by #171.

For me, it's actually about thinking ahead. If we're going to start adding state legislators and other possible contacts for issues, we're going to be seeing more complex cases around expected contacts. Properly handling and messaging why some contacts are included and others aren't for different issues will have a direct impact on how reliable and trustworthy our users see the app.

So at the moment, I think splitDistrict and PR #175 gets us close enough for (I'm guessing...analytics could tell us) the majority of users, so I'm fine deferring this until things get more complicated. I think this will become a higher priority as we add more contact options (but then we'll have more information and may decide on a different solution, given that new information)

**Comment by ColemanCollins** [comment]
Yeah, this was a proposal for a much more robust way of understanding expected vs intended contacts, something I think it completely sense to table until it actually becomes a more complicated need.

What about for the time being; one addition to #175, which only handles the "no contacts at all for an issue" error case: a non-obtrusive, semi-generic "upsell" message that's not tied to the issue's contacts (intended or otherwise), just about their splitDistrict state; eg something like the attached. 

The language here could stand to be be tightened up, but the "may be causing you" phrasing is intentionally vague enough to allow it to appear on any issue page for someone with a splitDistrict. 

Advocating for tacking it on small/italic 'aside' and at the end is because it doesn't impact core functionally at all, and therefore should be secondary to actually making calls to people we know you can call. Plus putting it next to the "X more people to call for this issue" made logical sense in that it's a value that may be incorrect because of your splitDistrict status.
<img width="784" alt="screenshot 2017-02-13 16 06 36" src="https://cloud.githubusercontent.com/assets/1831834/22905729/fe2ad3cc-f206-11e6-8014-fb52ea63ad26.png">
